pkg/runtime/processor/subscriber/subscriber.go

```diff
@@ -0,0 +1,372 @@
+	s.streamSubs = make(map[string][]*subscription.Subscription)
+}
+
+func (s *Subscriber) InitProgramaticSubscriptions(ctx context.Context) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	return s.initProgramaticSubscriptions(ctx)
+}
+
+func (s *Subscriber) reloadPubSubStream(name string, pubsub *rtpubsub.PubsubItem) error {
+	for _, sub := range s.streamSubs[name] {
+		sub.Stop()
+	}
+	s.streamSubs[name] = nil
+
+	if s.closed || pubsub == nil {
+		return nil
+	}
+
+	subs := make([]*subscription.Subscription, 0, len(s.compStore.ListSubscriptionsStreamByPubSub(name)))
+	var errs []error
+	for _, sub := range s.compStore.ListSubscriptionsStreamByPubSub(name) {
+		ss, err := subscription.New(subscription.Options{
+			AppID:           s.appID,
+			Namespace:       s.namespace,
+			PubSubName:      name,
+			Topic:           sub.Topic,
+			IsHTTP:          s.isHTTP,
+			PubSub:          pubsub,
+			Resiliency:      s.resiliency,
+			TraceSpec:       s.tracingSpec,
+			Route:           sub,
+			Channels:        s.channels,
+			GRPC:            s.grpc,
+			Adapter:         s.adapter,
+			AdapterStreamer: s.adapterStreamer,
+		})
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to create subscription for %s: %s", name, err))
+			continue
+		}
+
+		subs = append(subs, ss)
+	}
+
+	s.streamSubs[name] = subs
+
+	return errors.Join(errs...)
+}
+
+func (s *Subscriber) reloadPubSubApp(name string, pubsub *rtpubsub.PubsubItem) error {
+	for _, sub := range s.appSubs[name] {
+		sub.Stop()
+	}
+
+	s.appSubs[name] = nil
+
+	if !s.appSubActive || s.closed || pubsub == nil {
+		return nil
+	}
+
+	if err := s.initProgramaticSubscriptions(context.TODO()); err != nil {
+		return err
+	}
+
+	var errs []error
+	subs := make([]*subscription.Subscription, 0, len(s.compStore.ListSubscriptionsAppByPubSub(name)))
+	for _, sub := range s.compStore.ListSubscriptionsAppByPubSub(name) {
+		ss, err := subscription.New(subscription.Options{
+			AppID:      s.appID,
+			Namespace:  s.namespace,
+			PubSubName: name,
+			Topic:      sub.Topic,
+			IsHTTP:     s.isHTTP,
+			PubSub:     pubsub,
+			Resiliency: s.resiliency,
+			TraceSpec:  s.tracingSpec,
+			Route:      sub,
+			Channels:   s.channels,
+			GRPC:       s.grpc,
+			Adapter:    s.adapter,
+		})
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to create subscription for %s: %s", name, err))
+			continue
+		}
+
+		subs = append(subs, ss)
+	}
+
+	s.appSubs[name] = subs
+
+	return errors.Join(errs...)
+}
+
+func (s *Subscriber) initProgramaticSubscriptions(ctx context.Context) error {
+	if s.hasInitProg {
+		return nil
+	}
+
+	// If no pubsubs registered, return early.
+	if len(s.compStore.ListPubSubs()) == 0 {
+		return nil
+	}
+
+	appChannel := s.channels.AppChannel()
+	if appChannel == nil {
+		log.Warn("app channel not initialized, make sure -app-port is specified if pubsub subscription is required")
+		return nil
+	}
+
+	s.hasInitProg = true
```

Comment:
This means the method `initProgramaticSubscriptions` cannot be retried - this might be ok, just wanted to call out that's the case.

Reply:
As today, this function will continue to only be called once.